src/transformers/generation_utils.py

```diff
@@ -837,15 +887,34 @@ def sample(
             scores = logits_processor(input_ids, next_token_logits)
             scores = logits_warper(input_ids, scores)
 
-            # sample
-            probs = F.softmax(scores, dim=-1)
-            next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
-
-            # add code that transfomers next_tokens to tokens_to_add
+            if (cur_len == starting_length) and (num_return_sequences > 1) and diverse_sequences:
+                # sample num_return_sequences tokens
+                probs = F.softmax(scores, dim=-1)
+                next_tokens = torch.multinomial(probs, num_samples=num_return_sequences).reshape(-1)
+                # Once we got next_tokens, we have expand metadata
+                unfinished_sequences = unfinished_sequences.repeat_interleave(num_return_sequences, dim=0)
```

Comment:
Won't this quickly explode the number of current sequences?

Reply:
its purpose to explode sequences. Instead of exlploding them before calling `sample` and start each sequence from scratch (when nothing stops first token of each sequence to be equal (to the first token of another sequence)) this ensures first tokens to differ